### PR TITLE
Js5 project bean services apis

### PR DESCRIPTION
**Pull request by HarshaM0211**
Merged the code having Field Lables in to the develop branch, then created a new branch(this branch) from develop and merged the previous - Js5 project service- api branch.

### CODE REVIEW

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+        final User user = dao.read(User.class, userId, true);
+        final Criteria criteria = Criteria.equal(ProjectUser.PROP_USER, user);
+
+        final List<ProjectUser> userProjects = dao.find(ProjectUser.class, criteria, pageNo, pageSize);
+
+        final List<ProjectDTO> projectDtos = new ArrayList<ProjectDTO>();
+        for (ProjectUser pu : userProjects) {
+            final Project project = pu.getProject();
+            final ProjectDTO projectDto = this.toDTO(project);
```
Comment (VyshnaviRG) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java:223]: Null check

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+    // Criteria
+    // ------------------------------------------------------------------------
+
+    public Criteria getProjectUsersCriteria(final Integer projectId, final Integer userId) {
+
+        final User userEntity = dao.read(User.class, userId, true);
+        final Project projectEntity = dao.read(Project.class, projectId, true);
+
+        final Criteria projectcriteria = Criteria.equal(ProjectUser.PROP_PROJECT, projectEntity);
```
Comment (Vignesh1799) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java]: [PP] You can do static import of variables. `PROP_PROJECT`, `PROP_USER`

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        Verify.notNull(projectId, "$addUser :: projectId must be non NULL");
+
+        final ProjectUser projectUser = dao.findOne(ProjectUser.class, this.getProjectUsersCriteria(projectId, userId));
+
+        if (Objects.nonNull(projectUser)) {
+            LOGGER.info("User with Id : %s, already Exists in the specified Project with Id : %s", userId, projectId);
+            return;
+        }
+
+        final User userEntity = dao.read(User.class, userId, true);
```
Comment (Vignesh1799) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java]: The User and Project DB calls already happened inside `getProjectUsersCriteria(Integer projectId, Integer userId)`.
Instead make the method  `getProjectUsersCriteria(Project project, User user)` and make the DB Call once.

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        // Sanity Checks
+        Verify.notNull(projectId, "$update :: projectId must be non NULL");
+        Verify.notNull(projectBean, "$update :: projectBean must be non NULL");
+
+        super.update(projectId, projectBean);
+    }
+
+    @Override
+    @Transactional
+    public void addUser(final Integer userId, final Integer projectId) {
```
Comment (Vignesh1799) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java:167]: Create a separate service and implementation for the Project User table. In the Future, there might be a requirement for additional fields (column) which end up in additional services.
In that case, separate services make sense. we can debate on this if required.

File: src/main/java/com/mandark/jira/web/rest/ProjectAPI.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        final Pagination pagination = Pagination.with(count, pageNo, pageSize);
+        final PageResult pageResult = PageResult.with(pagination, projetDtos);
+
+        return new ResponseEntity<>(pageResult, HttpStatus.OK);
+    }
+
+    // Project :: Read the Projects by UserId
+    // ------------------------------------------------------------------------
+
+    @RequestMapping(value = "/users/{userId}", method = RequestMethod.GET)
```
Comment (Vignesh1799) [inline src/main/java/com/mandark/jira/web/rest/ProjectAPI.java:91]: We can debate on this API Path `api/v1/orgs/{orgId}/projects/users/{userId}` with @VyshnaviRG @Rupesh8844

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    // Private Methods for Key Generation
+    // ------------------------------------------------------------------------
+
+    private List<String> generateKeys(final Integer orgId, final String projectName) {
+
+        // Sanity Checks
+        Verify.notNull(projectName, "$generateKeys :: projectName must be non NULL");
+        Verify.notNull(orgId, "$generateKeys :: orgId must be non NULL");
```
Comment (VyshnaviRG) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java:297]: Please Check Verify.hasLength()

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+        final Organisation organisation = dao.read(Organisation.class, orgId, true);
+
+        Project project = new Project();
+        if (Objects.nonNull(entityBean.getProjectKey())) {
+            project = this.createFromBean(entityBean);
+        }
+        if (Objects.isNull(entityBean.getProjectKey())) {
+
+            List<String> keyList = this.generateKeys(orgId, entityBean.getName());
+            for (String key : keyList) {
```
Comment (VyshnaviRG) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java]: What if List is empty or you don't get any unique keys

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    // Private Methods for Key Generation
+    // ------------------------------------------------------------------------
+
+    private List<String> generateKeys(final Integer orgId, final String projectName) {
+
+        // Sanity Checks
+        Verify.notNull(projectName, "$generateKeys :: projectName must be non NULL");
+        Verify.notNull(orgId, "$generateKeys :: orgId must be non NULL");
+
+        final String[] subStrings = projectName.split(" ");
```
Comment (VyshnaviRG) [inline src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java:299]: What if projectName has just single/double letters

File: src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java (hunk 1 above)
Comment (HarshaM0211) [reply src/main/java/com/mandark/jira/app/service/impl/ProjectServiceImpl.java:223]: I made Project field in ProjectUser.java class as Non-Nullable

**Pull request merged**